fvm/meter/interaction_meter.go

```diff
@@ -51,30 +57,39 @@ func (m *InteractionMeter) MeterStorageRead(
 ) error {
 
 	// all reads are on a View which only read from storage at the first read of a given key
-	if _, ok := m.storageUpdateSizeMap[storageKey]; !ok {
+	if _, ok := m.reads[storageKey]; !ok {
 		readByteSize := getStorageKeyValueSize(storageKey, value)
 		m.totalStorageBytesRead += readByteSize
-		m.storageUpdateSizeMap[storageKey] = readByteSize
+		m.reads[storageKey] = readByteSize
 	}
 
 	return m.checkStorageInteractionLimit(enforceLimit)
 }
 
 // MeterStorageWrite captures storage written bytes count and returns an error
-// if it goes beyond the total interaction limit and limit is enforced
+// if it goes beyond the total interaction limit and limit is enforced.
+// If a key is written multiple times, only the last write is counted.
+// If a key is written before it has been read, next time it will be read it will be from the view,
```

Comment:
This behavior sounds like the metering module is making assumptions of the actual behavior of reading storage, such as the second time reading from storage will actually read from the view instead. 

However, can the metering module guarantee that the actual behavior matches with the assumption?

Wouldn't it be simplier of we don't need the metering to understand the behavior, but wraps the storage, so that when it's actually being read from storage, then it will be metered?

Reply:
That's is a very good point. Unfortunately I don't yet fully know how that would work.

The problem is with the programs cache. It stores a detached state that is attached when you get a program from the cache (to replay register reads).  Because of this, when reading a value from the state, you cannot know if that is going to be fetched from the parent state or from storage.

This is a bit complicated, so let me try to illustrate it better with an example.
Lets say that during loading of program *MyContract* you need to read register *R*. After this contract is first used in a transaction it will be in the cache, together with a *State* (a delta / a recording) of all registers touched during loading of *MyContract* including the read of register *R*.

When this contract will be used from the cache in future transaction there are 2 possibilities for register *R*:
- *R* was not touched (get-ed) before loading/applying *MyContract* from the programs cache, in which case *R* should be counted as being retrieved from storage
- *R* was touched (get-ed) before loading/applying *MyContract* from the programs cache, in which case *R* should be counted as being retrieved from the parent state

This logic is easier to do in meter.Merge then applying in State.Merge and then fixing the metered values from there.